### PR TITLE
chore!: adopt log/slog, drop go-kit/log

### PR DESCRIPTION
**Pull request by SuperQ**
This PR includes:
- linter updates to enable `sloglint` linter
- Go dep updates for prometheus/{client_golang,common,exporter-toolkit} libs
- refactorings to adopt log/slog in favor of go-kit/log

The bulk of this PR was automated by the following script which is being used to aid in converting the various exporters/projects to use slog:

https://gist.github.com/tjhop/49f96fb7ebbe55b12deee0b0312d8434

### CODE REVIEW

**Comment by SuperQ** [comment]
:eyeglasses: @tjhop

File: cmd/client/main.go [hunk 1]
```diff
@@ -89,11 +88,11 @@ func newBackOffFromFlags() backoff.BackOff {
 
 // Coordinator for scrape requests and responses
 type Coordinator struct {
-	logger log.Logger
+	logger *slog.Logger
 }
 
 func (c *Coordinator) handleErr(request *http.Request, client *http.Client, err error) {
-	level.Error(c.logger).Log("err", err)
+	c.logger.Error("coordinator error", "error", err)
```
Comment (dswarbrick) [inline cmd/client/main.go:95]:
```suggestion
	c.logger.Error("Coordinator error", "error", err)
```

File: cmd/client/main.go [hunk 2]
```diff
@@ -182,28 +181,28 @@ func (c *Coordinator) doPush(resp *http.Response, origRequest *http.Request, cli
 func (c *Coordinator) doPoll(client *http.Client) error {
 	base, err := url.Parse(*proxyURL)
 	if err != nil {
-		level.Error(c.logger).Log("msg", "Error parsing url:", "err", err)
+		c.logger.Error("Error parsing url:", "err", err)
```
Comment (dswarbrick) [inline cmd/client/main.go:184]:
```suggestion
		c.logger.Error("Error parsing URL:", "err", err)
```

File: cmd/client/main.go [hunk 3]
```diff
@@ -182,28 +181,28 @@ func (c *Coordinator) doPush(resp *http.Response, origRequest *http.Request, cli
...
 		return fmt.Errorf("error parsing url: %w", err)
 	}
 	u, err := url.Parse("poll")
 	if err != nil {
-		level.Error(c.logger).Log("msg", "Error parsing url:", "err", err)
+		c.logger.Error("Error parsing url:", "err", err)
```
Comment (dswarbrick) [inline cmd/client/main.go:189]:
```suggestion
		c.logger.Error("Error parsing URL:", "err", err)
```

File: cmd/client/main.go [hunk 4]
```diff
@@ -221,32 +220,32 @@ func (c *Coordinator) loop(bo backoff.BackOff, client *http.Client) {
 		if err := backoff.RetryNotify(op, bo, func(err error, _ time.Duration) {
 			pollErrorCounter.Inc()
 		}); err != nil {
-			level.Error(c.logger).Log("err", err)
+			c.logger.Error("backoff returned error", "error", err)
```
Comment (dswarbrick) [inline cmd/client/main.go:223]:
```suggestion
			c.logger.Error("Backoff returned error", "error", err)
```

File: cmd/client/main.go [hunk 5]
```diff
@@ -221,32 +220,32 @@ func (c *Coordinator) loop(bo backoff.BackOff, client *http.Client) {
...
+	flag.AddFlags(kingpin.CommandLine, &promslogConfig)
 	kingpin.HelpFlag.Short('h')
 	kingpin.Parse()
-	logger := promlog.New(&promlogConfig)
+	logger := promslog.New(&promslogConfig)
 	coordinator := Coordinator{logger: logger}
 
 	if *proxyURL == "" {
-		level.Error(coordinator.logger).Log("msg", "--proxy-url flag must be specified.")
+		coordinator.logger.Error("--proxy-url flag must be specified.")
```
Comment (dswarbrick) [inline cmd/client/main.go:237]:
```suggestion
		coordinator.logger.Error("--proxy-url flag must be specified")
```

File: cmd/client/main.go [hunk 6]
```diff
@@ -221,32 +220,32 @@ func (c *Coordinator) loop(bo backoff.BackOff, client *http.Client) {
...
 	*proxyURL = strings.TrimRight(*proxyURL, "/") + "/"
-	level.Info(coordinator.logger).Log("msg", "URL and FQDN info", "proxy_url", *proxyURL, "fqdn", *myFqdn)
+	coordinator.logger.Info("URL and FQDN info", "proxy_url", *proxyURL, "fqdn", *myFqdn)
 
 	tlsConfig := &tls.Config{}
 	if *tlsCert != "" {
 		cert, err := tls.LoadX509KeyPair(*tlsCert, *tlsKey)
 		if err != nil {
-			level.Error(coordinator.logger).Log("msg", "Certificate or Key is invalid", "err", err)
+			coordinator.logger.Error("Certificate or Key is invalid", "err", err)
```
Comment (dswarbrick) [inline cmd/client/main.go:248]:
```suggestion
			coordinator.logger.Error("Certificate or key is invalid", "err", err)
```

File: cmd/client/main.go [hunk 7]
```diff
@@ -257,12 +256,12 @@ func main() {
 	if *caCertFile != "" {
 		caCert, err := os.ReadFile(*caCertFile)
 		if err != nil {
-			level.Error(coordinator.logger).Log("msg", "Not able to read cacert file", "err", err)
+			coordinator.logger.Error("Not able to read cacert file", "err", err)
```
Comment (dswarbrick) [inline cmd/client/main.go:259]:
```suggestion
			coordinator.logger.Error("Unable to read cacert file", "err", err)
```

File: cmd/client/main.go [hunk 8]
```diff
@@ -257,12 +256,12 @@ func main() {
...
 			os.Exit(1)
 		}
 		caCertPool := x509.NewCertPool()
 		if ok := caCertPool.AppendCertsFromPEM(caCert); !ok {
-			level.Error(coordinator.logger).Log("msg", "Failed to use cacert file as ca certificate")
+			coordinator.logger.Error("Failed to use cacert file as ca certificate")
```
Comment (dswarbrick) [inline cmd/client/main.go:264]:
```suggestion
			coordinator.logger.Error("Failed to use cacert file as CA certificate")
```

File: cmd/proxy/main.go [hunk 9]
```diff
@@ -145,13 +144,13 @@ func (h *httpHandler) handlePoll(w http.ResponseWriter, r *http.Request) {
 	fqdn, _ := io.ReadAll(r.Body)
 	request, err := h.coordinator.WaitForScrapeInstruction(strings.TrimSpace(string(fqdn)))
 	if err != nil {
-		level.Info(h.logger).Log("msg", "Error WaitForScrapeInstruction:", "err", err)
+		h.logger.Info("Error WaitForScrapeInstruction:", "err", err)
```
Comment (dswarbrick) [inline cmd/proxy/main.go:147]: Should this be `h.logger.Error(...)`?

File: cmd/client/main.go [hunk 10]
```diff
@@ -272,7 +271,7 @@ func main() {
 	if *metricsAddr != "" {
 		go func() {
 			if err := http.ListenAndServe(*metricsAddr, promhttp.Handler()); err != nil {
-				level.Warn(coordinator.logger).Log("msg", "ListenAndServe", "err", err)
+				coordinator.logger.Warn("ListenAndServe", "err", err)
```
Comment (dswarbrick) [inline cmd/client/main.go:274]: Should this be `coordinator.logger.Error(...)`?

File: cmd/client/main.go [hunk 11]
```diff
@@ -137,13 +136,13 @@ func (c *Coordinator) doScrape(request *http.Request, client *http.Client) {
 		c.handleErr(request, client, fmt.Errorf("failed to scrape %s: %w", request.URL.String(), err))
 		return
 	}
-	level.Info(logger).Log("msg", "Retrieved scrape response")
+	logger.Info("Retrieved scrape response")
 	if err = c.doPush(scrapeResp, request, client); err != nil {
 		pushErrorCounter.Inc()
-		level.Warn(logger).Log("msg", "Failed to push scrape response:", "err", err)
+		logger.Warn("Failed to push scrape response:", "err", err)
```
Comment (dswarbrick) [inline cmd/client/main.go:142]: Should this be `logger.Error(...)`?

File: cmd/client/main.go [hunk 12]
```diff
@@ -102,14 +101,14 @@ func (c *Coordinator) handleErr(request *http.Request, client *http.Client, err
 	}
 	if err = c.doPush(resp, request, client); err != nil {
 		pushErrorCounter.Inc()
-		level.Warn(c.logger).Log("msg", "Failed to push failed scrape response:", "err", err)
+		c.logger.Warn("Failed to push failed scrape response:", "err", err)
```
Comment (dswarbrick) [inline cmd/client/main.go:104]: Should this be `c.logger.Error(...)`?

**Review by dswarbrick — COMMENTED** [review]
Overall mostly ok, just a few nits here and there. Some of the logging calls that pass an error as a KV should perhaps actually be `logger.Error(...)` instead of `Warn` or `Info`. I think some of the `Info` logging calls are also actually more like `Debug`.

I'm not sure how necessary it is to log an error immediately before returning a wrapped error, e.g.
```go
		c.logger.Error("Error polling:", "err", err)
		return fmt.Errorf("error polling: %w", err)
```
In cases like that, it probably makes more sense that the caller chooses whether to log the returned error.

**Comment by SuperQ** [comment]
Decisions about what to log when, as well as changing a lot of the logging strings, is scope creep. Additional improvements in logging are something that will be done in separate PRs.

**Pull request merged**